### ISSUE TITLE
[P1] Is this repo support MPT architecture ? i got error

### ISSUE THREAD

**Issue by srn-source** [stanfordnlp/pyreft#40]
![image](https://github.com/stanfordnlp/pyreft/assets/107049099/db6795b7-fefb-42d4-aa99-57f262896036)
I am running model Instruction finetuning with MPT architecture but i got this error as image. when i changed to GPT2. don't have any problem. Is this repo support MPT architecture ?

Please suggest me.

**Comment by frankaging** [comment]
@srn-source hey, thanks for raising the issue. could you provide your `reft_config`?

i think the thing to try here is to use string access for MPT in your `reft_config`! don't use "block_output", that component is LLaMA-specific.

**Comment by srn-source** [comment]
@frankaging  I use defaults  from this tutorial without changing. and i am new for this method.
This is config.
https://github.com/stanfordnlp/pyreft/blob/f9f83384a876350d9eee86ccad9f6a921e2044c6/examples/loreft/train.py#L276

**Comment by frankaging** [comment]
@srn-source got it. thanks.

one example will be: let say we are using this MPT model from huggingface:

 https://github.com/huggingface/transformers/blob/v4.39.3/src/transformers/models/mpt/modeling_mpt.py#L512

and let's say we want to intervene on some middle layer residual stream, then you can try something like

```py
        representations = [{
            "component": f"transformer.block[{l}].output",
            "low_rank_dimension": rank,
            "intervention": intervention_type(
                embed_dim=config.hidden_size, low_rank_dimension=rank,
                dropout=dropout, dtype=intervention_dtype, act_fn=act_fn, device=device,
                add_bias=add_bias
            )
        } for l in layers]
```

note that here, you are specifying the component via, what we call it, **string access**. `transformer.block[5].output` is the string access to the intervening model component.

You can also check what we did for `RoBERTa` models. It is following the same way,
https://github.com/stanfordnlp/pyreft/blob/main/examples/loreft/train.py#L267

Let me know if it is working! Ty.

**Comment by srn-source** [comment]
@frankaging  Wow its working ..... but my model does not support Left Padding https://github.com/stanfordnlp/pyreft/blob/f9f83384a876350d9eee86ccad9f6a921e2044c6/examples/loreft/dataset.py#L220

is it possible to add a single padding token AFTER input_ids?

**Comment by frankaging** [comment]
@srn-source the step here is not for left padding. (we might clarify this later with a structured document).

ReFT training requires a special `sinking token` on the left. This `sinking token` will **take care of intervening sequences in a batch with different sequence length.** For instance, in LoReFT, we do prefix and suffix interventions, e.g., intervening on the first 3 and last 3 tokens for every sequence. As a result, for sequences with different lengths, the last 3 tokens' position indices will differ. We use the first special token to take care that.

Our dataset loader will add a pad to the front, and also adjust the attention mask accordingly. 

You can see that the tokenizer is still right padding:
https://github.com/stanfordnlp/pyreft/blob/f9f83384a876350d9eee86ccad9f6a921e2044c6/examples/loreft/train.py#L149

Let me know if this makes sense. You can also just try to do some Alpaca-training with your MPT model, and see if it is working?

Note that, during inference, you also need to provide the intervening location like in our demo:
https://github.com/stanfordnlp/pyreft/blob/f9f83384a876350d9eee86ccad9f6a921e2044c6/examples/chat/chat_model.ipynb

**Comment by frankaging** [comment]
Closing this issue as the documentation as well as dataset class cleaning effort will be tracked here:
https://github.com/stanfordnlp/pyreft/issues/43

Thanks again for raising the issue.
